ios/browser/api/ai_chat/ai_chat_conversation_driver.mm

```diff
@@ -0,0 +1,124 @@
+#include "brave/components/ai_chat/core/common/mojom/ai_chat.mojom-shared.h"
+#include "brave/components/ai_chat/core/common/mojom/ai_chat.mojom.h"
+#include "brave/ios/browser/api/ai_chat/ai_chat_delegate.h"
+#include "brave/ios/browser/skus/skus_service_factory.h"
+#include "components/user_prefs/user_prefs.h"
+#include "ios/chrome/browser/shared/model/application_context/application_context.h"
+#include "ios/chrome/browser/shared/model/browser_state/chrome_browser_state.h"
+#include "ios/chrome/common/channel_info.h"
+#include "net/base/mac/url_conversions.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "url/gurl.h"
+
+namespace ai_chat {
+
+AIChatConversationDriverIOS::AIChatConversationDriverIOS(
+    id<AIChatDelegate> delegate,
+    raw_ptr<ChromeBrowserState> browser_state,
+    raw_ptr<AIChatMetrics> ai_chat_metrics,
+    PrefService* local_state_prefs)
+    : ConversationDriver(user_prefs::UserPrefs::Get(browser_state),
+                         ai_chat_metrics,
+                         std::make_unique<AIChatCredentialManager>(
+                             GetSkusService(browser_state),
+                             local_state_prefs),
+                         std::make_unique<ai_chat::AIChatFeedbackAPI>(
+                             browser_state->GetSharedURLLoaderFactory(),
+                             GetChannelString()),
+                         browser_state->GetSharedURLLoaderFactory()),
+      bridge_(delegate) {
+  chat_driver_observation_.Observe(this);
+}
+
+AIChatConversationDriverIOS::~AIChatConversationDriverIOS() = default;
+
+std::u16string AIChatConversationDriverIOS::GetPageTitle() const {
+  NSString* title = [bridge_ getPageTitle];
+  return title ? base::SysNSStringToUTF16(title) : std::u16string();
+}
+
+GURL AIChatConversationDriverIOS::GetPageURL() const {
+  return net::GURLWithNSURL([bridge_ getLastCommittedURL]);
+}
+
+void AIChatConversationDriverIOS::GetPageContent(
+    base::OnceCallback<void(std::string, bool is_video)> callback) const {
+  [bridge_
+      getPageContentWithCompletion:[callback =
+                                        std::make_shared<decltype(callback)>(
+                                            std::move(callback))](
+                                       NSString* content, bool isVideo) {
+        if (callback) {
+          std::move(*callback).Run(
+              content ? base::SysNSStringToUTF8(content) : std::string(),
+              isVideo);
+        }
+      }];
+}
+
+bool AIChatConversationDriverIOS::HasPrimaryMainFrame() const {
+  return [bridge_ hasPrimaryMainFrame];
+}
+
+bool AIChatConversationDriverIOS::IsDocumentOnLoadCompletedInPrimaryMainFrame()
+    const {
+  return [bridge_ isDocumentOnLoadCompletedInPrimaryFrame];
+}
+
+base::RepeatingCallback<mojo::PendingRemote<skus::mojom::SkusService>()>
+AIChatConversationDriverIOS::GetSkusService(ChromeBrowserState* browser_state) {
+  return base::BindRepeating(
```

Comment:
passing in the browser state and using it to get other services from the factory is an anti-pattern in chromium. You should pass in the skus service, prefs and shared url loader factory into the `AIChatConversationDriverIOS`, but I'm concerned about lifetimes here, is an `AIChatConversationDriver` explicitly tied to a tab in ios and are all tabs destroyed before we shutdown the chromium services?

Reply:
Fixed. Renamed the class. VPN and all classes that use Skus pass in a getter. Refactoring it would mean I have to refactor Desktop and Android again.